### PR TITLE
AL/BE-99 Changes administrator functions.

### PR DESCRIPTION
**Pull request by lut1k**

### CODE REVIEW

File: pallets/controller/src/lib.rs [hunk 1]
```diff
@@ -392,25 +501,33 @@ impl<T: Trait> Module<T> {
...
+						sum_borrow_plus_effects,
+						borrow_amount,
+						oracle_price,
+					)?;
+				}
 			}
 		}
 
+		ensure!(
+			!((sum_collateral == sum_borrow_plus_effects) && (sum_collateral == 0)),
```
Comment (Pavel-Blaize) [inline pallets/controller/src/lib.rs:527]: I am not sure if we need this check and assert. Zero collateral and zero borrow_sum can be a legal situation, especially when this function is called in view/calculation purposes. All other checks for 0 amounts are already performed on the upper levels.

File: pallets/controller/src/lib.rs [hunk 2]
```diff
@@ -362,6 +463,10 @@ impl<T: Trait> Module<T> {
 		// For each tokens the account is in
 		for asset in m_tokens_ids.into_iter() {
 			let underlying_asset = Self::get_underlying_asset_id_by_wrapped_id(&asset)?;
+			if !<LiquidityPools<T>>::check_user_available_collateral(&account, underlying_asset) {
```
Comment (Pavel-Blaize) [inline pallets/controller/src/lib.rs:466]: Problem in this check - user now is not able to get borrow in the currency which is not a collateral.

File: pallets/controller/src/lib.rs [hunk 3]
```diff
@@ -362,6 +463,10 @@ impl<T: Trait> Module<T> {
...
+				continue;
+			}
+
 			let m_token_balance = T::MultiCurrency::free_balance(asset, account);
 			if m_token_balance == Balance::zero() {
```
Comment (Pavel-Blaize) [inline pallets/controller/src/lib.rs:471]: Problem in this check - the user is not able to take a bottow in a currency which is not deposited

**Review by Pavel-Blaize — APPROVED** [review]

File: pallets/controller/src/lib.rs (hunk 1 above)
Comment (lut1k) [reply pallets/controller/src/lib.rs]: Removed this check.

File: pallets/controller/src/lib.rs (hunk 2 above)
Comment (lut1k) [reply pallets/controller/src/lib.rs]: Fixed.

File: pallets/controller/src/lib.rs (hunk 3 above)
Comment (lut1k) [reply pallets/controller/src/lib.rs]: Fixed.

File: pallets/controller/src/lib.rs [hunk 4]
```diff
@@ -362,52 +463,66 @@ impl<T: Trait> Module<T> {
...
-			let m_token_balance = T::MultiCurrency::free_balance(asset, account);
-			if m_token_balance == Balance::zero() {
-				continue;
-			}
 
-			// Read the balances and exchange rate from the cToken
-			let borrow_balance = Self::borrow_balance_stored(account, underlying_asset)?;
-			let exchange_rate = Self::get_exchange_rate(underlying_asset)?;
-			let collateral_factor = Self::get_collateral_factor(underlying_asset);
+			if <LiquidityPools<T>>::check_user_available_collateral(&account, underlying_asset) {
```
Comment (Pavel-Blaize) [inline pallets/controller/src/lib.rs:467]: There is still a scenario, when the user tries to take a borrow in a currency which is not marked as available for collateral, and he fails.
```
for asset in m_tokens_ids.into_iter() {
        let underlying_asset = Self::get_underlying_asset_id_by_wrapped_id(&asset)?;
			
         // Get the normalized price of the asset.
        let oracle_price = <Oracle...;
            
        if oracle_price == Price::zero() {
		return Ok((Balance::zero(), Balance::zero()));
	}
        let exchange_rate = Self::get_exchange_rate(underlying_asset)?;
	let collateral_factor = Self::get_collateral_factor(underlying_asset);
         // Pre-compute a conversion factor from tokens -> dollars (normalized price value)
	let tokens_to_denom = ...
        
        if <LiquidityPools<T>>::check_user_available_collateral(&account, underlying_asset) {
		let m_token_balance = T::MultiCurrency::free_balance(asset, account);
		let borrow_balance = ....
               if m_token_balance != Balance::zero() {
			// sum_collateral += tokens_to_denom * m_token_balance
                       // sum_borrow_plus_effects += oracle_price * borrow_balance
               }
        }
	// Calculate effects of interacting with Underlying Asset Modify
	if underlying_to_borrow == underlying_asset {
		// redeem effect
		// borrow effect
	}
}
```

**Review by Pavel-Blaize — DISMISSED** [review]

File: pallets/controller/src/lib.rs (hunk 4 above)
Comment (lut1k) [reply pallets/controller/src/lib.rs]: Fixed.

File: pallets/controller/src/lib.rs [hunk 5]
```diff
@@ -376,38 +475,56 @@ impl<T: Trait> Module<T> {
...
-			sum_collateral =
-				Self::mul_price_and_balance_add_to_prev_value(sum_collateral, m_token_balance, tokens_to_denom)?;
+			if <LiquidityPools<T>>::check_user_available_collateral(&account, underlying_asset) {
+				let m_token_balance = T::MultiCurrency::free_balance(asset, account);
 
-			// sum_borrow_plus_effects += oracle_price * borrow_balance
-			sum_borrow_plus_effects =
-				Self::mul_price_and_balance_add_to_prev_value(sum_borrow_plus_effects, borrow_balance, oracle_price)?;
+				// Pre-compute a conversion factor from tokens -> dollars (normalized price value)
+				tokens_to_denom = collateral_factor
```
Comment (Pavel-Blaize) [inline pallets/controller/src/lib.rs:486]: tokens_to_denom value is also used in redeem effect in the 'if' block on line 509. For now it uses default value or the value from the previous cycle step. Should be calculated in the beginning of the cycle

Comment (lut1k) [reply pallets/controller/src/lib.rs]: Fixed. Check please.

**Review by Pavel-Blaize — APPROVED** [review]

**Review by Apollo50 — APPROVED** [review]

**Pull request merged**